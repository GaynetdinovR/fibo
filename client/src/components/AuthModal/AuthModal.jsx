import styles from "../../styles/components/AuthModal.module.sass";

import Modal from "../../ui/Modal.jsx";
import H3 from "../../ui/H3.jsx";
import Button from "../../ui/Button.jsx";

import PhoneNumberInput from "./components/PhoneNumberInput.jsx";
import CodeInput from "./components/CodeInput.jsx";
import BottomSide from "./components/BottomSide.jsx";

import { NotificationManager } from "react-notifications";
import { useReducer } from "react";
import { useDispatch } from "react-redux";
import { login } from "../../store/userSlice/userSlice.js";

import { formatPhone } from "../../scripts/functions.js";
import { authorization } from "../../scripts/api.js";

const reducer = (state, action) => {
	switch (action.type) {
		case "switch_disabled_phone":
			return { ...state, isDisabledPhoneInput: !state.isDisabledPhoneInput };
		case "set_code_sent":
			return { ...state, isCodeSent: action.val };
		case "set_disabled_phone":
			return { ...state, isDisabledPhoneInput: action.val };
		case "set_code_errored":
			return { ...state, isCodeErrored: action.val };
		case "set_phone_errored":
			return { ...state, isPhoneErrored: action.val };
		case "set_code":
			return { ...state, code: action.val };
		case "set_code_input_val":
			return { ...state, codeInputVal: action.val };
		case "set_phone_input_val":
			return { ...state, phoneNumberInputVal: action.val };
		default:
			throw new Error("Unknown action: " + action.type);
	}
};

const AuthModal = ({ setOpen, isOpen }) => {
	const initState = {
		phoneNumberInputVal: "",
		codeInputVal: "",
		code: "0000",
		isCodeSent: false,
		isDisabledPhoneInput: false,
		isPhoneErrored: false,
		isCodeErrored: false
	};

	const [authModalState, localDispatch] = useReducer(reducer, initState);

	const setValToAuthState = (type, val) => {
		localDispatch({ type: type, val: val });
	};

	const dispatch = useDispatch();

	/**
	 * Действия после нажатия кнопки "Войти"
	 */
	const logInBtnClicked = async () => {
		if (authModalState.codeInputVal !== authModalState.code)
			return setValToAuthState("set_code_errored", true);

		const phoneNumber = formatPhone(authModalState.phoneNumberInputVal);

		dispatch(login(phoneNumber));
		NotificationManager.success("Вы успешно зашли в аккаунт");
		setOpen(false);

		await authorization(phoneNumber);
	};

	//Elements

	const logInBtn = (
		<Button
			onClickFn={logInBtnClicked}
			className={styles.auth_modal__log_in_btn}
		>
			Войти
		</Button>
	);

	return (
		<Modal className={styles.auth_modal} setOpen={setOpen} isOpen={isOpen}>
			<div className={styles.auth_modal__content}>
				<H3 className={styles.auth_modal__title}>Вход на сайт</H3>

				<PhoneNumberInput
					setValToAuthState={setValToAuthState}
					authModalState={authModalState}
				/>

				{authModalState.isCodeSent ? (
					<CodeInput
						authModalState={authModalState}
						setValToAuthState={setValToAuthState}
					/>
				) : null}

				{!authModalState.isCodeSent ? (
					<BottomSide
						authModalState={authModalState}
						setValToAuthState={setValToAuthState}
					/>
				) : null}

				{authModalState.isCodeSent ? logInBtn : null}
			</div>
		</Modal>
	);
};

export default AuthModal;
