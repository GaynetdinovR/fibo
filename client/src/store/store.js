import { configureStore } from "@reduxjs/toolkit";
import userReducer from "./userSlice/userSlice.js";
import productsReducer from "./productsSlice/productsSlice.js";

const store = configureStore({
	reducer: {
		user: userReducer,
		products: productsReducer
	}
});

export default store;
