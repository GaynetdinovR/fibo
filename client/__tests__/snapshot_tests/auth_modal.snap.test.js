import renderer from "react-test-renderer";
import { it, expect } from "@jest/globals";
import AuthModal from "../../src/components/AuthModal/AuthModal.jsx";
import { Provider } from "react-redux";
import store from "../../src/store/store.js";

let state = true;
const set = (val) => {
	state = val;
};

it("renders correctly", () => {
	const tree = renderer
		.create(
			<Provider store={store}>
				<AuthModal isOpen={state} setOpen={set} />
			</Provider>
		)
		.toJSON();
	expect(tree).toMatchSnapshot();
});
