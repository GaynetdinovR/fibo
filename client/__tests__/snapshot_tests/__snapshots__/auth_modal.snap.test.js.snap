// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="modal__wrapper"
  onClick={[Function]}
>
  <aside
    className="auth_modal modal"
    onClick={[Function]}
  >
    <button
      className="modal__close_btn"
      onClick={[Function]}
    >
      <img
        alt="close"
        src="./icons/cross.png"
      />
    </button>
    <div
      className="auth_modal__content"
    >
      <h4
        className="auth_modal__title h3"
      >
        Вход на сайт
      </h4>
      <label
        className="auth_modal__input_wrap"
        onClick={[Function]}
      >
        <span
          className="auth_modal__input_label"
        >
          Номер телефона
        </span>
        <div
          className="input__wrap"
        >
          <input
            className="auth_modal__input input"
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onMouseDown={[Function]}
            onPaste={[Function]}
            placeholder="+7 999 999-99-99"
          />
          <div
            className="none"
          >
            <div
              className="input_error_alert__img"
            >
              <img
                alt="warn"
                src="./icons/warn.png"
              />
            </div>
            <span
              className="input_error_alert__text"
            >
              Неправильный вид номера
            </span>
          </div>
        </div>
      </label>
      <div
        className="auth_modal__bottom_side"
      >
        <button
          className="auth_modal__get_code_btn default_button"
          onClick={[Function]}
        >
          Выслать код
        </button>
        <p
          className="auth_modal__terms_of_use text"
        >
          <span>
            Продолжая, вы соглашаетесь 
          </span>
          <a
            href="https://en.wikipedia.org/wiki/Terms_of_service"
            target="blank"
          >
            со сбором и обработкой персональных данных и пользовательским соглашением
          </a>
        </p>
      </div>
    </div>
  </aside>
</div>
`;
